collect_app/src/test/java/org/odk/collect/android/savepoints/SavepointFinderTest.kt

```diff
@@ -0,0 +1,316 @@
+            ),
+            equalTo(savepoint2)
+        )
+    }
+
+    @Test
+    fun `getSavepoint returns null if savepoint exists in the database but the file does not`() {
+        val savepointFile = createSavepointFile()
+        val savepoint = Savepoint(formV2.dbId, null, savepointFile.absolutePath, "")
+        savepointsRepository.save(savepoint)
+        savepointFile.delete()
+
+        assertThat(
+            SavepointFinder.getSavepoint(
+                FormsContract.getUri("1", formV2.dbId),
+                FormsContract.CONTENT_ITEM_TYPE,
+                formsRepository,
+                instancesRepository,
+                savepointsRepository
+            ),
+            equalTo(null)
+        )
+    }
+
+    @Test
+    fun `getSavepoint called for a saved form uri returns null if only its blank form has a savepoint`() {
+        val savepointFile = createSavepointFile()
+        val savepoint = Savepoint(formV2.dbId, null, savepointFile.absolutePath, instance1.instanceFilePath)
+        savepointsRepository.save(savepoint)
+
+        assertThat(
+            SavepointFinder.getSavepoint(
+                InstancesContract.getUri("1", instance1.dbId),
+                InstancesContract.CONTENT_ITEM_TYPE,
+                formsRepository,
+                instancesRepository,
+                savepointsRepository
+            ),
+            equalTo(null)
+        )
+    }
+
+    @Test
+    fun `getSavepoint called for a saved form uri returns savepoint if it has one`() {
+        val savepointFile = createSavepointFile()
+        val savepoint = Savepoint(formV2.dbId, instance1.dbId, savepointFile.absolutePath, instance1.instanceFilePath)
+        savepointsRepository.save(savepoint)
+
+        assertThat(
+            SavepointFinder.getSavepoint(
+                InstancesContract.getUri("1", instance1.dbId),
+                InstancesContract.CONTENT_ITEM_TYPE,
+                formsRepository,
+                instancesRepository,
+                savepointsRepository
+            ),
+            equalTo(savepoint)
+        )
+    }
+
+    @Test
+    fun `getSavepoint called for a saved form uri returns null if savepoint exists in the database but the file does not`() {
+        val savepointFile = createSavepointFile()
+        val savepoint = Savepoint(formV2.dbId, instance1.dbId, savepointFile.absolutePath, instance1.instanceFilePath)
+        savepointsRepository.save(savepoint)
+        savepointFile.delete()
+
+        assertThat(
+            SavepointFinder.getSavepoint(
+                InstancesContract.getUri("1", instance1.dbId),
+                InstancesContract.CONTENT_ITEM_TYPE,
+                formsRepository,
+                instancesRepository,
+                savepointsRepository
+            ),
+            equalTo(null)
+        )
+    }
+
+    @Test
+    fun `getSavepoint called for a saved form uri returns null if it has one but the instance file has been modified later`() {
```

Comment:
Is the only time this can happen if an instance is updated on disk (not through Collect)? Otherwise we'd get a new save which deletes the save point right?

Either way, could we use `instance#lastStatusChangeDate` as a comparison instead? That changes on every save as far as I can see (it's badly named). That would allow you to avoid the problems with file creation times.

Reply:
> Is the only time this can happen if an instance is updated on disk (not through Collect)? Otherwise we'd get a new save which deletes the save point right?

Yeah, I think so. I keep it just in case but I belive it shouldn't be needed.

> Either way, could we use instance#lastStatusChangeDate as a comparison instead? That changes on every save as far as I can see (it's badly named). That would allow you to avoid the problems with file creation times.

Done.